pandas/_libs/tslibs/conversion.pyx

```diff
@@ -390,10 +411,22 @@ cdef _TSObject create_tsobject_tz_using_offset(npy_datetimestruct dts,
         check_overflows(obj)
         return obj
 
+    # Infer fold from offset-adjusted obj.value
+    if is_utc(tz) or is_tzlocal(tz):
+        # TODO: think on how we can infer fold for local Timezone
+        # and adjust value for fold
+        pass
+    else:
+        trans, deltas, typ = get_dst_info(tz)
+
+        if typ in ['pytz', 'dateutil']:
+            pos = trans.searchsorted(obj.value, side='right') - 1
+            fold = _infer_tsobject_fold(obj, trans, deltas, pos)
+
     # Keep the converter same as PyDateTime's
     dt = datetime(obj.dts.year, obj.dts.month, obj.dts.day,
                   obj.dts.hour, obj.dts.min, obj.dts.sec,
-                  obj.dts.us, obj.tzinfo)
+                  obj.dts.us, obj.tzinfo, fold=fold)
```

Comment:
`obj.tzinfo` is a `pytz.FixedOffset`. `fold` is always `0`, so you can remove this and the "infer tsobject fold" logic above.

Reply:
@pganssle 
The problem is that then you call `convert_datetime_to_tsobject`, and if you haven't set or inferred `dt.fold`, then you end up losing the ability to infer from unambiguous string.

This is what happens if we cut this argument and logic:

```
================================== FAILURES ===================================
_ test_timestamp_constructor_infer_fold_from_value[2019-10-27 01:30:00+00:00-1-Europe/London] _

tz = 'Europe/London', ts_input = '2019-10-27 01:30:00+00:00', fold_out = 1

    @pytest.mark.parametrize("tz", ["dateutil/Europe/London", "Europe/London"])
    @pytest.mark.parametrize(
        "ts_input,fold_out",
        [
            (1572136200000000000, 0),
            (1572139800000000000, 1),
            ("2019-10-27 01:30:00+01:00", 0),
            ("2019-10-27 01:30:00+00:00", 1),
            (datetime(2019, 10, 27, 1, 30, 0, 0, fold=0), 0),
            (datetime(2019, 10, 27, 1, 30, 0, 0, fold=1), 1),
        ],
    )
    def test_timestamp_constructor_infer_fold_from_value(tz, ts_input, fold_out):
        # Test for #25057
        # Check that we infer fold correctly based on timestamps since utc
        # or strings
        ts = pd.Timestamp(ts_input, tz=tz)
        result = ts.fold
        expected = fold_out
>       assert result == expected
E       assert 0 == 1
```

Reply:
Again, `pytz` does not support `fold`, so if you remove it from the matrix, you shold be OK here.

I think you may be misunderstanding what `fold` is supposed to be. `fold` solves a specific problem: there is not enough information contained in a `datetime.datetime` object without fold to distinguish between ambiguous datetimes. This is because `tzinfo` is a set of functions that takes the naive portion of the `datetime` and maps it to information about the offset. This doesn't work when two datetimes are distinguished *only by their offset*, so an additional bit is added to the naive portion of the `datetime` to allow time zone providers to uniquely map "naive portion" to offsets.

`pytz` *does not have this problem*, because `pytz` worked around this problem by inventing its own semantics for this. It ends up storing the relevant information in the time zone itself.

You only need to set the value for `fold` if your time zone provider needs it, and you would never expect `pytz` to set it. You can throw away all tests that involve `pytz` and `fold`.

Reply:
should we then default fold=None and if it is passed and we have a pytz zone raise?

Reply:
@pganssle 
Thank you. This is what I was having trouble with, because I could see that Timestamps based on pytz produced different offsets for different sides of fold.
Now that I understand your explanation, I understand how this is done:

```python
        elif typ in ['pytz', 'dateutil']:
            pos = trans.searchsorted(obj.value, side='right') - 1
            if typ == 'pytz':
                tz = tz._tzinfos[tz._transition_info[pos]]
            dt64_to_dtstruct(obj.value + deltas[pos], &obj.dts)
```

Basically, we search through `tz._tzinfos` and manually pick the timezone that looks like what we need. This thing is completely separate from the `fold` mechanism.

I agree that adjusting or inferring fold for pytz is meaningless and we should just raise if fold is passed with a pytz-like timezone.

Reply:
> should we then default fold=None and if it is passed and we have a pytz zone raise?

I think I'm fine with raising in this situation, but I will say that this is a further deviation from `datetime.datetime`'s behavior, which may have somewhat unpredictable consequences.

One possible downside is that if `pytz` ever changes to support `fold`, you'll have to change the behavior of `fold` in `pandas`, and `pytz` users who want a fold-supporting `pytz` in `pandas` will need to pull in the latest `pandas`.

It's unlikely to happen and I think `pandas` is already OK with an uncomfortably tight coupling with its dependencies like this, but it is just one potential pitfall.

That said, the permissions ratchet usually goes one way, so maybe it's best to start by raising and then remove it if people come up with ways that it fails in real code.

Reply:
This logic should still be removed, though, right? You're constructing a `datetime` that always has a `pytz.FixedOffset` for its `tzinfo`, so I expect that `fold` will always be ignored.

Reply:
@pganssle 
No, it gets set: datetime seems to be very permissive about setting fold. This allows us to pass the `fold` we've inferred to `convert_datetime_to_tsobject` through `dt`, which uses the `fold` attribute from its `datetime` input to build `_TSObject`.